internal/utils/utils.go

```diff
@@ -12,24 +13,33 @@ import (
 	"google.golang.org/protobuf/proto"
 )
 
-// NonceLength is the length of the string returned by NewNonce. NewNonce encodes the current UNIX
-// time in nanos in hex encoding, so the nonce will be 16 characters if the current UNIX nano time is
-// greater than 2^60-1. This is because it takes 16 hex characters to encode 64 bits, but only 15 to
-// encode 60 bits (the output of strconv.FormatInt is not padded by 0s). 2^60-1 nanos from epoch time
-// (January 1st 1970) is 2006-07-14 23:58:24.606, which as of this writing is over 17 years ago. This
-// is why it's guaranteed that NonceLength will be 16 characters (before that date, encoding the
-// nanos only required 15 characters). For the curious, the UNIX nano timestamp will overflow int64
-// some time in 2262, making this constant valid for the next few centuries.
-const NonceLength = 16
-
-// NewNonce creates a new unique nonce based on the current UNIX time in nanos. It always returns a
-// string of length NonceLength.
-func NewNonce() string {
-	// The second parameter to FormatInt is the base, e.g. 2 will return binary, 8 will return octal
-	// encoding, etc. 16 means FormatInt returns the integer in hex encoding, e.g. 30 => "1e" or
-	// 1704239351400 => "18ccc94c668".
-	const hexBase = 16
-	return strconv.FormatInt(time.Now().UnixNano(), hexBase)
+const (
+	// NonceLength is the length of the string returned by NewNonce. NewNonce encodes the current UNIX
+	// time in nanos and the remaining chunks, encoded as 64-bit and 32-bit integers respectively, then
+	// hex encoded. This means a nonce will always be 8 + 4 bytes, multiplied by 2 by the hex encoding.
+	NonceLength = (8 + 4) * 2
+)
+
+// NewNonce creates a new unique nonce based on the current UNIX time in nanos, always returning a
+// string of [NonceLength].
+func NewNonce(remainingChunks int) string {
```

Comment:
a smart yet big change to the nonce. To make sure it's backward compatible, I checked Rest.li xds client is not using the nonce in any way currently. May want to let Xin and Yan know to make sure Envoy is not using it too.

Reply:
No one is, it's not supposed to be used for anything other than ACKs/NACKs, it's not meant to be interpreted more than that. Envoy doesn't use it at all for sure.